Mathlib/NumberTheory/FLT/Polynomial.lean

```diff
@@ -0,0 +1,337 @@
+/-
+Copyright (c) 2024 Jineon Baek and Seewoo Lee. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jineon Baek, Seewoo Lee
+-/
+import Mathlib.Algebra.Polynomial.Expand
+import Mathlib.NumberTheory.FLT.Basic
+import Mathlib.NumberTheory.FLT.MasonStothers
+import Mathlib.RingTheory.Polynomial.Content
+
+/-!
+# Fermat's Last Theorem for polynomials over a field
+
+This file states and proves the Fermat's Last Theorem for polynomials over a field.
+For `n ≥ 3` not divisible by the characteristic of the coefficient field `k` and (pairwise) nonzero
+coprime polynomials `a, b, c` (over a field) with `a ^ n + b ^ n = c ^ n`,
+all polynomials must be constants.
+
+More generally, we can prove non-solvability of Fermat-Catalan equation: there are no
+non-constant polynomial solution of the equation `u * a ^ p + v * b ^ q + w * c ^ r = 0`, where
+`p, q, r ≥ 3` with `p * q + q * r + r * p ≤ p * q * r` and not divisible by `char k`
+and `u, v, w` are nonzero elements in `k`.
+
+The proof uses Mason-Stothers theorem (Polynomial ABC theorem) and infinite descent
+(for characteristic p case).
+-/
+
+open Polynomial UniqueFactorizationMonoid UniqueFactorizationDomain
+
+variable {k : Type _} [Field k]
+variable {R : Type _} [CommRing R] [IsDomain R] [NormalizationMonoid R]
+  [UniqueFactorizationMonoid R]
+
+private lemma Ne.isUnit_C {u : k} (hu : u ≠ 0) : IsUnit (C u) :=
+  Polynomial.isUnit_C.mpr hu.isUnit
+
+-- Multiplying units preserve coprimality
+private lemma isCoprime_mul_units_left {a b : k[X]} {u v : k} (hu : u ≠ 0) (hv : v ≠ 0) :
```

Comment:
@jcpaik It seems that if we replace `u ≠ 0` into `IsUnit u` for auxiliary lemmas (not main theorems) then proofs could be a bit simpler (for main theorems, keep as nonzero assumption and use `isUnit_C` for example). How do you think? At least for this `isCoprime_mul_units_left`, I'll use the unit version PR.

Reply:
I found that some part of the proofs use both of $u \ne 0$ and $u$ being a unit (to be precise, `C u` being nonzero or unit), so replacing this does not make proofs significantly shorter, so I only modified a few. (`Ne.isUnit_C` is useful enough)